pyro/infer/util.py

```diff
@@ -124,68 +125,71 @@ def __repr__(self):
             '({}, ...)'.format(frames) for frames in self]))
 
 
-class TreeSum(object):
+def _dict_iadd(items, key, value):
+    if key in items:
+        items[key] = items[key] + value
+    else:
+        items[key] = value
+
+
+class MultiFrameDice(object):
     """
-    Data structure to compute cumulative costs along paths in a tree.
-    Typically keys are ``cond_indep_stack``s.
+    An implementation of the DiCE operator compatible with Pyro features.
+
+    This implementation correctly handles:
+    - scaled log-probability due to subsampling
+    - independence in different contexts due to iarange
+    - weights due to parallel and sequential enumeration
+
+    This assumes restricted dependency structure on the model and guide:
```

Comment:
"This assumes restricted dependency..."

are we going to add checks for this?

Reply:
We can add some partial checks before release, but some cases are impossible to detect. E.g. in
```py
x_axis = iarange('x')
y_axis = iarange('y')
with x_axis:
    x = pyro.sample('x', ...)
with y_axis:
    y = pyro.sample('y', ...)
with x_axis, y_axis:
    xy = pyro.sample("xy", ...)
```
it is impossible to determine whether `y` depends on `x`. (we could examine the autograd graph, but the user could get around this via `x.detach()`).